### PR TITLE
[SPARK-11781][SPARKR] SparkR has problem in inferring type of raw type.

### PR DESCRIPTION
**Pull request by sun-rui**

### CODE REVIEW

**Comment by shivaram** [comment]
Could you describe the problem a bit more ? Was it that raw vectors were being treated as lists ? Just curious how raw vectors differ from int vectors etc.

**Comment by sun-rui** [comment]
@shivaram, The R raw type is intended to hold raw bytes. while int vector is to hold 32-bit integer values. The R raw type maps to Spark SQL binary type, which is internally represented in Array[Byte].

This PR solves two problems:
1. Inferring of raw type is incorrect. 
```
> SparkR:::infer_type(as.raw(c(1, 2 ,3)))
[1] "array<binary>"
```
This is not correct, it should be "binary".

2. Collecting a DataFrame fails if there is any column of binary type. The bug lies in the logic that determines whether a collected column can be coerced into a atomic vector or not.

**Comment by shivaram** [comment]
Ok thanks for the clarification. It might take me a couple of days to get to this as the change looks a bit involved.

cc @felixcheung

File: R/pkg/R/DataFrame.R [hunk 1]
```diff
@@ -717,25 +717,30 @@ setMethod("collect",
...
+                # Get a cell from a column of complex type returns a list instead of a vector.
                 col <- listCols[[colIndex]]
                 if (length(col) <= 0) {
                   df[[names[colIndex]]] <- col
                 } else {
                   # TODO: more robust check on column of primitive types
-                  vec <- do.call(c, col)
-                  if (class(vec) != "list") {
-                    df[[names[colIndex]]] <- vec
+                  if (!any(sapply(col, function(e) { length(e) > 1 }))) {
```
Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: should this be returned from org.apache.spark.sql.api.r.SQLUtils.dfToCols for more optimal processing? this seems like potentially a lot of data to go through

**Comment by sun-rui** [comment]
@felixcheung, you concern is reasonable. I refactor the code by using schema to determine if a collected column can be coerced into an atomic vector.

**Comment by sun-rui** [comment]
@shivaram, @felixcheung, could you take more look? another PR depends on this one.

File: R/pkg/R/DataFrame.R [hunk 2]
```diff
@@ -700,25 +700,28 @@ setMethod("collect",
...
+                  df[[colName]] <- col
                 } else {
-                  # TODO: more robust check on column of primitive types
-                  vec <- do.call(c, col)
-                  if (class(vec) != "list") {
-                    df[[names[colIndex]]] <- vec
+                  colType <- dtypes[[colIndex]][[2]]
+                  if (!is.null(PRIMITIVE_TYPES[[colType]]) && colType != "binary") {
+                    vec <- do.call(c, col)
+                    stopifnot (class(vec) != "list")
```
Comment (felixcheung) [inline R/pkg/R/DataFrame.R]: nit: no space for func call: `stopifnot(class(vec) != "list")`

**Comment by felixcheung** [comment]
looks good.

File: R/pkg/R/DataFrame.R (hunk 2 above)
Comment (sun-rui) [inline R/pkg/R/DataFrame.R]: fixed

File: R/pkg/R/DataFrame.R [hunk 3]
```diff
@@ -700,25 +700,28 @@ setMethod("collect",
...
                 if (length(col) <= 0) {
-                  df[[names[colIndex]]] <- col
+                  df[[colName]] <- col
                 } else {
-                  # TODO: more robust check on column of primitive types
-                  vec <- do.call(c, col)
-                  if (class(vec) != "list") {
-                    df[[names[colIndex]]] <- vec
+                  colType <- dtypes[[colIndex]][[2]]
+                  if (!is.null(PRIMITIVE_TYPES[[colType]]) && colType != "binary") {
```
Comment (shivaram) [inline R/pkg/R/DataFrame.R]: Could you add a comment here as well ? Something like `NOTE: "binary" columns behave like complex types`

**Comment by shivaram** [comment]
LGTM but for a minor comment.

File: R/pkg/R/DataFrame.R (hunk 3 above)
Comment (sun-rui) [inline R/pkg/R/DataFrame.R]: added

**Comment by shivaram** [comment]
LGTM. Merging this

**Pull request closed without merging**